pkg/tun/tun_linux.go

```diff
@@ -0,0 +1,153 @@
+// tun_linux.go: Open a Tunnel (L3 virtual interface) using the Universal TUN/TAP device driver.
+package tun
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"os"
+	"unsafe"
+
+	"golang.org/x/sys/unix"
+
+	"github.com/datawire/dlib/dexec"
+	"github.com/telepresenceio/telepresence/v2/pkg/tun/buffer"
+)
+
+const devicePath = "/dev/net/tun"
+
+type Device struct {
+	*os.File
+	name  string
+	index uint32
+}
+
+// OpenTun creates a new TUN device and ensures that it is up and running.
+func OpenTun() (*Device, error) {
+	// https://www.kernel.org/doc/Documentation/networking/tuntap.txt
+
+	fd, err := unix.Open(devicePath, unix.O_RDWR, 0)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open TUN device %s: %v", devicePath, err)
+	}
+	unix.CloseOnExec(fd)
+	defer func() {
+		if err != nil {
+			_ = unix.Close(fd)
+		}
+	}()
+
+	var flagsRequest struct {
+		name  [unix.IFNAMSIZ]byte
+		flags int16
+	}
+	copy(flagsRequest.name[:], "tel%d")
+	flagsRequest.flags = unix.IFF_TUN | unix.IFF_NO_PI
+
+	err = unix.IoctlSetInt(fd, unix.TUNSETIFF, int(uintptr(unsafe.Pointer(&flagsRequest))))
+	if err != nil {
+		return nil, err
+	}
+
+	var name string
+	for i := 0; i < unix.IFNAMSIZ; i++ {
+		if flagsRequest.name[i] == 0 {
+			name = string(flagsRequest.name[0:i])
+			break
+		}
+	}
+	if name == "" {
+		name = string(flagsRequest.name[:])
+	}
+
+	htons := func(value uint16) uint16 {
+		test := uint16(1)
+		if (*[2]byte)(unsafe.Pointer(&test))[0] == 1 {
+			// this machine is little endian, swap bytes
+			value = value&0xff<<8 | value&0xff00>>8
+		}
+		return value
+	}
+
+	// Passing a network ordered short here is required.
+	provisioningSocket, err := unix.Socket(unix.AF_PACKET, unix.SOCK_RAW, int(htons(unix.ETH_P_ALL)))
+	if err != nil {
+		return nil, err
+	}
+	defer unix.Close(provisioningSocket)
+
+	if err = unix.BindToDevice(provisioningSocket, name); err != nil {
+		return nil, err
+	}
```

Comment:
Where's this come from?

If I look at net-tools `ifconfig` (the usual `ifconfig` on Linux, unless you're using BusyBox `ifconfig), it uses
`socket(AF_INET, SOCK_DGRAM, IPPROTO_IP)` and then doesn't bother binding it to a device.

Can we get a comment with explanation or references or something?

Reply:
I should have made a comment because now I cannot find the link. I think I got it from the ifconfig source but perhaps it was the wrong one. In any case, using the `socket()` parameters that you suggest and removing the `BindToDevice()` call works just fine, so I'll do that.